src/main/java/org/jitsi/videobridge/cc/vp8/VP8AdaptiveTrackProjectionContext.java

```diff
@@ -125,330 +96,393 @@ public VP8AdaptiveTrackProjectionContext(
         {
-            int previousMaxSequenceNumber
-                = getMaxSequenceNumberOfFrame(ssrc, timestamp);
+            /* Leave a gap for the unfinished end of the previously routed frame. */
+            projectedSeqGap++;
 
-            if (previousMaxSequenceNumber != -1
-                && RtpUtils.isOlderSequenceNumberThan(
-                    previousMaxSequenceNumber, sequenceNumber))
-            {
-                frameToMaxSequenceNumberMap.put(timestamp, sequenceNumber);
-            }
+            /* Make sure subsequent packets of the previous projection won't
+               overlap the new one.
+             */
+            lastVP8FrameProjection.close();
+        }
+
+        int projectedSeq = RtpUtils.applySequenceNumberDelta(lastVP8FrameProjection.getLatestProjectedSequence(), projectedSeqGap);
+
+        // this is a simulcast switch. The typical incremental value =
+        // 90kHz / 30 = 90,000Hz / 30 = 3000 per frame or per 33ms
+        long tsDelta;
+        if (lastVP8FrameProjection.getCreatedMs() != 0)
+        {
+            tsDelta = 3000 * Math.max(1, (nowMs - lastVP8FrameProjection.getCreatedMs()) / 33);
         }
         else
         {
-            frameToMaxSequenceNumberMap.put(timestamp, sequenceNumber);
+            tsDelta = 3000;
         }
+        long projectedTs = RtpUtils.applyTimestampDelta(lastVP8FrameProjection.getTimestamp(), tsDelta);
+
+        int picId;
+        int tl0PicIdx;
+        if (lastVP8FrameProjection.getVP8Frame() != null)
+        {
+            picId = Vp8Utils.applyPictureIdDelta(lastVP8FrameProjection.getPictureId(),
+                1);
+            tl0PicIdx = Vp8Utils.applyTl0PicIdxDelta(lastVP8FrameProjection.getTl0PICIDX(),
+                1);
+        }
+        else {
+            picId = frame.getPictureId();
+            tl0PicIdx = frame.getTl0PICIDX();
+        }
+
+        VP8FrameProjection projection =
+            new VP8FrameProjection(diagnosticContext, logger,
+                frame, lastVP8FrameProjection.getSSRC(), projectedTs,
+                RtpUtils.getSequenceNumberDelta(projectedSeq, initialPacket.getSequenceNumber()),
+                picId, tl0PicIdx, nowMs
+            );
+
+        return projection;
     }
 
     /**
-     * @return true if this instance needs a keyframe, false otherwise.
+     * Create a projection for this frame. It is being sent subsequent to other projected frames
+     * of this layer.
      */
-    @Override
-    public boolean needsKeyframe()
+    @NotNull
+    private VP8FrameProjection createInLayerProjection(@NotNull VP8Frame frame, @NotNull VP8Frame refFrame, @NotNull Vp8Packet initialPacket)
     {
-        if (vp8QualityFilter.needsKeyframe())
-        {
-            logger.debug(() -> "quality filter "
-                    + vp8QualityFilter.hashCode()
-                    + " says vp8 track needs keyframe");
+        long nowMs = System.currentTimeMillis();
```

Comment:
Don't you have tests fo these? How do they work if they're using clock time?

Reply:
Yeah, this is slightly dodgy; it works by the fact that the minimum time difference it'll handle is 33 ms, which is more than the time the test should take to run.  There's a comment in the unit test that the test case will fail if you stop it in a debugger at the wrong moment.